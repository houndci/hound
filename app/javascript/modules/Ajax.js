import $ from "jquery";

export function createSubscription(subscriptionOptions) {
  return $.ajax({
    url: `/repos/${subscriptionOptions.repo_id}/subscription.json`,
    type: "POST",
    data: subscriptionOptions,
    dataType: "json"
  });
}

export function deleteSubscription(repo) {
  return $.ajax({
    url: `/repos/${repo.id}/subscription`,
    type: "DELETE",
    dataType: "json"
  });
}

export function deactivateRepo(repo) {
  return $.ajax({
    url: `/repos/${repo.id}/deactivation`,
    type: "POST",
    dataType: "text"
  });
}

export function activateRepo(repo) {
  return $.ajax({
    url: `/repos/${repo.id}/activation`,
    type: "POST",
    dataType: "text"
  });
}

export function updateOwner(id, params) {
  return $.ajax({
    url: `/owners/${id}`,
    type: "PUT",
    dataType: "json",
    data: { owner: params }
  });
}

export function upgradeSubscription(id, params) {
  return $.ajax({
    url: `/repos/${id}/subscription.json`,
    type: "PUT",
    dataType: "json",
    data: params,
    success: () => {
      document.location.href = "/repos";
    }
  });
}
