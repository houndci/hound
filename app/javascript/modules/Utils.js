export function getCSRFfromHead() {
  if (process.env.NODE_ENV === 'test') {
    return "csrf_token";
  } else {
    return document.querySelector("meta[name=csrf-token]").content;
  }
}
