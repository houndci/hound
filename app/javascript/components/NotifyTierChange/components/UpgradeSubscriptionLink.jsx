import React from 'react';
import $ from 'jquery';

import * as Ajax from '../../../modules/Ajax';
import { getCSRFfromHead } from '../../../modules/Utils';

export default class UpgradeSubscriptionLink extends React.Component {
  componentWillMount() {
    $.ajaxSetup({
      headers: {
        "X-XSRF-Token": getCSRFfromHead()
      }
    });
  }

  upgradeWithNewCard(token) {
    const { repoId } = this.props;
    Ajax.upgradeSubscription(repoId, { card_token: token.id });
  }

  checkout() {
    const { stripePublishableKey, iconPath } = Hound.settings;

    return StripeCheckout.configure({
      key: stripePublishableKey,
      image: iconPath,
      token: (token) => this.upgradeWithNewCard(token),
    });
  }

  showCreditCardForm() {
    const { userEmailAddress } = Hound.settings;
    const { price, title } = this.props.nextTier;

    this.checkout().open({
      name: title,
      amount: price * 100,
      email: userEmailAddress,
      panelLabel: "{{amount}} per month",
      allowRememberMe: false
    });
  }

  handleClick() {
    const { repoId, userHasCard } = this.props;

    if (userHasCard) {
      Ajax.upgradeSubscription(repoId);
    } else {
      this.showCreditCardForm();
    }
  }

  render() {
    return(
      <a
        className="repo-toggle tier-change-accept"
        href="javascript:void(0);"
        onClick={() => this.handleClick()}
      >Upgrade</a>
    );
  }
}
