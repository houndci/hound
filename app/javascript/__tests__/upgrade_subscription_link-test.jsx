import UpgradeSubscriptionLink from '../components/NotifyTierChange/components/UpgradeSubscriptionLink';
import * as Ajax from '../modules/Ajax';

it('renders appropriately', () => {
  const repo = {
    id: 1,
    name: "Test repo",
    owner: {
      id: 1
    }
  };

  const wrapper = shallow(
    <UpgradeSubscriptionLink
      authenticityToken={"csrf_token"}
      repoId={repo.id}
    />
  );
  expect(wrapper).toMatchSnapshot();
});


it('renders appropriately', () => {
  const ajaxSpy = sinon.spy(Ajax, 'upgradeSubscription');
  const nextTier = { price: 49, title: "Chihuahua" };

  const wrapper = mount(
    <UpgradeSubscriptionLink
      authenticityToken={"csrf_token"}
      nextTier={nextTier}
      repoId={1}
      userHasCard={true}
    />
  );

  wrapper.find('.repo-toggle').simulate('click');

  expect(ajaxSpy.calledOnce).toBe(true);
  expect(ajaxSpy.calledWith(1)).toBe(true);
});
