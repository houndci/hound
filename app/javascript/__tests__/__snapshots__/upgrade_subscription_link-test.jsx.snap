// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders appropriately 1`] = `
<a
  className="repo-toggle tier-change-accept"
  href="javascript:void(0);"
  onClick={[Function]}
>
  Upgrade
</a>
`;
